app/controllers/users/omniauth_callbacks_controller.rb

```diff
@@ -0,0 +1,106 @@
+class Users::OmniauthCallbacksController < Devise::OmniauthCallbacksController
+  before_filter :require_valid_omniauth, only: :nyulibraries
+  def nyulibraries
+    @user = find_user_with_or_without_provider.first_or_initialize(attributes_from_omniauth)
+    @user.update_attributes(attributes_from_omniauth)
+    @user.admin = true if is_admin?
+    @user.save!
+    if @user.persisted?
+      sign_in_and_redirect @user, event: :authentication
+      set_flash_message(:notice, :success, kind: "NYU Libraries") #if is_navigational_format?
+    else
+      session["devise.nyulibraries_data"] = request.env["omniauth.auth"]
+      redirect_to root_path
+    end
+  end
+
+  def find_user_with_or_without_provider
+    @find_user_with_or_without_provider ||= (find_user_with_provider.present?) ? find_user_with_provider : find_user_without_provider
+  end
+
+  def find_user_with_provider
+    @find_user_with_provider ||= User.where(username: omniauth.uid, provider: omniauth.provider)
+  end
+
+  def find_user_without_provider
+    @find_user_without_provider ||= User.where(username: omniauth.uid, provider: "")
+  end
+
+  def is_admin?
+    (Figs.env["MARLI_DEFAULT_ADMINS"] or []).include?(@user.username)
+  end
+
+  def require_valid_omniauth
+    head :bad_request unless valid_omniauth?
+  end
+
+  def valid_omniauth?
+    omniauth.present? && omniauth.provider.to_s == 'nyulibraries'
+  end
+
+  def omniauth
+    @omniauth ||= request.env["omniauth.auth"]
+  end
+
+  def omniauth_provider
+    @omniauth_provider ||= omniauth.provider
+  end
+
+  def user_attributes_from_aleph
+    omniauth_aleph_properties
+  end
+
+  def omniauth_aleph_properties
+    omniauth_aleph_identity.properties unless omniauth_aleph_identity.blank?
+  end
+
+  def attributes_from_omniauth
+    {
+      email: omniauth_email,
+      firstname: omniauth_firstname,
+      lastname: omniauth_lastname,
+      institution_code: omniauth_institution,
+      aleph_id: omniauth_aleph_id,
+      provider:  omniauth.provider
```

Comment:
@hab278 Make sure you populate the serialized address fields from Aleph. Use this as an example:

https://github.com/NYULibraries/marli/blob/master/app/models/user_session.rb#L11